Python/ast_opt.c

```diff
@@ -714,6 +721,125 @@ astfold_withitem(withitem_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
     return 1;
 }
 
+static int
+astfold_pattern_negative(expr_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
+{
+    assert(node_->kind == UnaryOp_kind);
+    assert(node_->v.UnaryOp.op == USub);
+    assert(node_->v.UnaryOp.operand->kind == Constant_kind);
+    PyObject *value = node_->v.UnaryOp.operand->v.Constant.value;
+    assert(PyComplex_CheckExact(value) ||
+           PyFloat_CheckExact(value) ||
+           PyLong_CheckExact(value));
+    PyObject *negated = PyNumber_Negative(value);
+    if (!negated) {
+        return 0;
+    }
+    assert(PyComplex_CheckExact(negated) ||
+           PyFloat_CheckExact(negated) ||
+           PyLong_CheckExact(negated));
+    return make_const(node_, negated, ctx_);
+}
+
+static int
+astfold_pattern_complex(expr_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
+{
+    expr_ty left = node_->v.BinOp.left;
+    expr_ty right = node_->v.BinOp.right;
+    if (left->kind == UnaryOp_kind) {
+        CALL(astfold_pattern_negative, expr_ty, left);
+    }
+    assert(left->kind = Constant_kind);
+    assert(right->kind = Constant_kind);
+    // LHS must be real, RHS must be imaginary:
+    if (!(PyFloat_CheckExact(left->v.Constant.value) ||
+          PyLong_CheckExact(left->v.Constant.value)) ||
+        !PyComplex_CheckExact(right->v.Constant.value))
+    {
+        // Not actually valid, but it's the compiler's job to complain:
+        return 1;
+    }
+    PyObject *new;
+    if (node_->v.BinOp.op == Add) {
+        new = PyNumber_Add(left->v.Constant.value, right->v.Constant.value);
+    }
+    else {
+        assert(node_->v.BinOp.op == Sub);
+        new = PyNumber_Subtract(left->v.Constant.value, right->v.Constant.value);
+    }
+    if (!new) {
+        return 0;
+    }
+    assert(PyComplex_CheckExact(new));
+    return make_const(node_, new, ctx_);
+}
+
+static int
+astfold_pattern_keyword(keyword_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
+{
+    CALL(astfold_pattern, expr_ty, node_->value);
+    return 1;
+}
+
+static int
+astfold_pattern(expr_ty node_, PyArena *ctx_, _PyASTOptimizeState *state)
+{
+    // Don't blindly optimize the pattern as an expr; it plays by its own rules!
+    // Currently, this is only used to form complex/negative numeric constants.
+    switch (node_->kind) {
```

Comment:
Something I noticed while working on PEP 642's reference implementation is that this ast_opt code is currently relying on the parser to keep the operators that it doesn't handle away from it, so it's easy to segfault by feeding it your own AST tree, without invoking the parser.

While filling in `validate_pattern` in ast.c will be the main answer to that, I think there are some genuinely missing cases here as well (e.g. `+1` will fail the UnaryOp check).

Reply:
> While filling in validate_pattern in ast.c will be the main answer to that, I think there are some genuinely missing cases here as well (e.g. +1 will fail the UnaryOp check).

The [drafted AST validator](https://github.com/isidentical/cpython/tree/patma-ast-validators) should prevent `+1`, but I can't say for sure is whether a full coverage exists (I tried to implement it by looking at the grammar but it would be nice if @brandtbucher can do a second pass on it): 
```
>>> t = ast.parse("match test:\n\tcase -1: pass")
>>> t.body[0].cases[0].pattern.op = ast.UAdd()
>>> print(ast.unparse(t))
match test:
    case +1:
        pass
>>> compile(t, "<test>", "exec")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
ValueError: only USub operator is allowed for UnaryOp nodes inside of a pattern
```

Reply:
> Something I noticed while working on PEP 642's reference implementation is that this ast_opt code is currently relying on the parser to keep the operators that it doesn't handle away from it, so it's easy to segfault by feeding it your own AST tree, without invoking the parser.

I'm not sure it's easy to actually segfault the compiler with malformed patterns; most of the assumptions it makes here and in `compile.c` about what the parser can emit are guarded either by `assert(...)` or `Py_UNREACHABLE()` calls. (I'm probably a heavier user of `assert` than the historical authors of these files). As I see it, the goal of `PyAST_Validate` is to make sure these checks cannot be hit by people playing with `ast` and `compile` in userland.

So checking those assumptions should be sufficient (and more should be added if missing).

> While filling in `validate_pattern` in ast.c will be the main answer to that, I think there are some genuinely missing cases here as well (e.g. `+1` will fail the UnaryOp check).

For what it's worth, `+1` isn't valid. Full grammar [here](https://www.python.org/dev/peps/pep-0634/#appendix-a-full-grammar).

> I can't say for sure is whether a full coverage exists (I tried to implement it by looking at the grammar but it would be nice if @brandtbucher can do a second pass on it)

Ha! I knew I was forgetting something. I promised you a review, then promptly got distracted with `ceval.c` and forgot. Thanks for reminding me.

Reply:
@isidentical can you open up a PR in your own repo (merging `isidentical/cpython:patma-ast-validators` into `isidentical/cpython:master`)? That way I can leave review comments.

Reply:
isidentical/cpython#1 should work

Reply:
Great, thanks!